### PR TITLE
feat: general implementation

### LINKED ISSUE #1: Make an authorization plugin for GraphQL
See: https://casbin.org/docs/en/graphql-middlewares but this one is already deprecated. We may need to maintain our own now.

image

We may borrow some ideas from this one: https://github.com/maticzav/graphql-shield#features

This issue tries to help the issue from Galoy: https://github.com/GaloyMoney/galoy/issues/294

image

### PR DESCRIPTION
**Pull request by Zxilly**
Fix: https://github.com/node-casbin/graphql-authz/issues/1

Signed-off-by: Zxilly <zhouxinyu1001@gmail.com>

### CODE REVIEW

**Comment by Zxilly** [comment]
@hsluoyz @Gabriel-403 plz review

**Comment by Sefriol** [comment]
Looks good to me, but:

I would like to enforce strict git rules before merge:

1. Squash unnecessary commits. Single feature should have a singular commit, fix commits are irrevelant when feature is implemented for the first time.
2. Other way around: single commit should include only things relevant to that commit.
3. Each commit should have single line description on what it's trying to do, followed by one empty line and then a longer description that gives more information about the given commit
    - For example, if you implement tests in singular commit, first line could be just a "Implement tests" and the description would be a summary of what tests were implemented

These rules would require following changes:

These commits should be combined into one:
https://github.com/node-casbin/graphql-authz/pull/2/commits/521bf3234656077b1cf4ebcafdc612dcdd7173e2
https://github.com/node-casbin/graphql-authz/pull/2/commits/2fc235b6dc94f6cc2365833b408be8c583df3881
https://github.com/node-casbin/graphql-authz/pull/2/commits/cff3f4cfbe2fca2dfc44e4ce7de92bf9c580f67e
https://github.com/node-casbin/graphql-authz/pull/2/commits/84f16832b42b9f7cb3e3d1bc8362786be629f2cf

Commit https://github.com/node-casbin/graphql-authz/pull/2/commits/521bf3234656077b1cf4ebcafdc612dcdd7173e2 should be split so that changes in `src/middleware.ts` are in https://github.com/node-casbin/graphql-authz/pull/2/commits/b3ce4011be5076c2720c835ac801c7172ad328f6

Also: I would like to have function descriptions included in `src/middleware.ts` before merge (input, output, functionality).

**Comment by Zxilly** [comment]
@Sefriol Thanks for review, working on this.

**Review by Sefriol — APPROVED** [review]
lgtm

**Review by nodece — APPROVED** [review]

**Comment by nodece** [comment]
You should add installation, example and so on to README.md.

File: tests/server.test.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
+import { applyMiddleware } from 'graphql-middleware';
```
Comment (Sefriol) [inline tests/server.test.ts:1]: Putting tests under github actions is probably a good idea before merge. Separate this into different commit.

File: src/directive.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+          type: GraphQLString,
+          defaultValue: '*',
+        },
+      },
+    });
+  }
+
+  public visitFieldDefinition(field: GraphQLField<any, any>): GraphQLField<any, any> | void | null {
+    const { resolve = defaultFieldResolver } = field;
+    // if (!(field as any).hasCanDirective) {
```
Comment (Sefriol) [inline src/directive.ts]: Remove commented lines before merge.

File: src/directive.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      } else {
+        if (role === '*') {
+          role = 'anonymous';
+        }
+        throw new PermissionInsufficientError(`${role} can not ${action} ${path}`);
+      }
+    };
+  }
+
+  // public visitObject(object: GraphQLObjectType): GraphQLObjectType | void | null {}
```
Comment (Sefriol) [inline src/directive.ts]: Remove commented lines before merge or add TODO, if something is missing.

File: src/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import { IMiddleware } from 'graphql-middleware';
+import { buildPathCache, CasbinAuthTypeName, CasbinContextPathCacheKey, CasbinContextRoleKey, Info2Path, PathAndAction2Key } from './utils';
+import { PermissionInsufficientError, RuleNotFoundError } from './error';
+import { GraphQLNonNull, GraphQLResolveInfo, GraphQLScalarType } from 'graphql';
+
+export class MiddlewareOptions {
+  enforcer: Enforcer;
+  ctxMember = 'role';
+}
+
```
Comment (Sefriol) [inline src/middleware.ts:28]: Add function description. (Input, output, describe what function does)

File: src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { GraphQLResolveInfo } from 'graphql';
+import { Enforcer } from 'casbin';
+
+export const CasbinContextPathCacheKey = '_casbinPathCache';
+export const CasbinContextEnforcerKey = '_casbinEnforcer';
+export const CasbinContextRoleKey = '_casbinRole';
+export const CasbinAuthTypeName = 'CasbinCanVisit';
+
+export const Info2Path = (info: GraphQLResolveInfo): string => {
+  let node = info.path;
```
Comment (Sefriol) [inline src/utils.ts:28]: Add function description. (Input, output, describe what function does)

File: src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    }
+  }
+  return fullPath;
+};
+
+export const PathAndAction2Key = (path: string, action: string): string => {
+  return `${path}_${action}`;
+};
+
+export const buildPathCache = async (CasbinInstance: Enforcer): Promise<Array<string>> => {
```
Comment (Sefriol) [inline src/utils.ts:52]: Add function description. (Input, output, describe what function does)

File: src/directive.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  CasbinContextEnforcerKey,
+  CasbinContextPathCacheKey,
+  CasbinContextRoleKey,
+  Info2Path,
+  PathAndAction2Key,
+} from './utils';
+import { EnforcerNotFoundError, PermissionInsufficientError } from './error';
+import { Enforcer } from 'casbin';
+
+export class CanDirective extends SchemaDirectiveVisitor {
```
Comment (Sefriol) [inline src/directive.ts:42]: Add class description.

File: src/directive.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  static getDirectiveDeclaration(directiveName: string, schema: GraphQLSchema): GraphQLDirective | null | undefined {
```
Comment (Sefriol) [inline src/directive.ts:49]: Add function description. (Input, output, describe what function does)

File: README.md [hunk 9]
```diff
@@ -1 +1,5 @@
-# graphql-authz
\ No newline at end of file
+# graphql-authz
```
Comment (Sefriol) [inline README.md:1]: Add a simple code example on how the library can be used.

File: src/directive.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        throw new PermissionInsufficientError(`${role} can not ${action} ${path}`);
+      }
+    };
+  }
+
+  // public visitObject(object: GraphQLObjectType): GraphQLObjectType | void | null {}
+}
+
+export class CanVisitType extends GraphQLScalarType {
```
Comment (Sefriol) [inline src/directive.ts:114]: Add Class description.

**Review by Sefriol — COMMENTED** [review]
Looks really good so far. I would add some modifications before merge.

Also remember to squash commits when needed. Refactor commits are not necessary in before merge.

File: src/directive.ts (hunk 3 above)
Comment (Zxilly) [reply src/directive.ts]: Yes, this is a pending work. I will add TODO

**Review by Sefriol — APPROVED** [review]
I did not run the tests, but otherwise lgtm.

**Comment by Zxilly** [comment]
@hsluoyz plz review

File: .gitignore [hunk 11]
```diff
@@ -1,104 +1,83 @@
...
-
-# TernJS port file
 .tern-port
+.vscode-test
+.yarn/cache
+.yarn/unplugged
+.yarn/build-state.yml
+.yarn/install-state.gz
+.pnp.*
+.idea/**/workspace.xml
```
Comment (hsluoyz) [inline .gitignore]: Just ignore .idea folder with one line

File: src/directive.ts [hunk 12]
```diff
@@ -0,0 +1,115 @@
+import { SchemaDirectiveVisitor } from '@graphql-tools/utils';
```
Comment (hsluoyz) [inline src/directive.ts:15]: License header for every file.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by Sefriol** [comment]
Squash these commits before merge.

Also, does apache2 license require adding license header to each file? @hsluoyz
